### PR TITLE
[CALCITE-5241] Implement CHAR function

### PR DESCRIPTION
**Pull request by JiajunBernoulli**
Notes: `char(n)` is different from `chr(n)` when n large than 256. `chr(n)` be added in https://issues.apache.org/jira/browse/CALCITE-2510

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -462,6 +462,16 @@ public static ByteString right(ByteString s, int n) {
     return s.substring(len - n);
   }
 
+  /**
+   * SQL CHAR(int) function.
+   */
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:467]: explain the difference between this and `chr`. It isn't int vs long.

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:467]: also make the javadoc consistent with other javadoc

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 2]
```diff
@@ -650,7 +650,16 @@ private SqlLibraryOperators() {
...
+  @LibraryOperator(libraries = {MYSQL, SPARK})
+  public static final SqlFunction CHAR =
+      new SqlFunction("CHAR",
+          SqlKind.OTHER_FUNCTION,
+          ReturnTypes.CHAR_FORCE_NULLABLE,
+          null,
+          OperandTypes.INTEGER,
+          SqlFunctionCategory.STRING);
+
+  @LibraryOperator(libraries = {ORACLE, POSTGRESQL})
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:662]: CHR and CHAR need javadoc

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 3]
```diff
@@ -347,6 +347,13 @@ public static SqlCall stripSeparator(SqlCall call) {
   public static final SqlReturnTypeInference CHAR =
           explicit(SqlTypeName.CHAR);
 
+  /**
+   * Type-inference strategy whereby the result type of a call is a nullable
+   * Char.
+   */
+  public static final SqlReturnTypeInference CHAR_FORCE_NULLABLE =
+      CHAR.andThen(SqlTypeTransforms.FORCE_NULLABLE);
+
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:356]: not 'nullable Char', but 'nullable CHAR(1)'

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 4]
```diff
@@ -1492,9 +1492,6 @@ protected static Calendar getCalendarNotTooNear(int timeUnit) {
     f.checkScalar("{fn ASCII('ABC')}", "65", "INTEGER NOT NULL");
     f.checkNull("{fn ASCII(cast(null as varchar(1)))}");
 
-    if (false) {
-      f.checkScalar("{fn CHAR(code)}", null, "");
-    }
```
Comment (julianhyde) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:1497]: Why remove this test? The goal of the change is to implement JDBC '{fn CHAR}'

File: core/src/main/java/org/apache/calcite/util/BuiltInMethod.java [hunk 5]
```diff
@@ -339,6 +339,7 @@ public enum BuiltInMethod {
   UPPER(SqlFunctions.class, "upper", String.class),
   LOWER(SqlFunctions.class, "lower", String.class),
   ASCII(SqlFunctions.class, "ascii", String.class),
+  CHAR(SqlFunctions.class, "charN", int.class),
   REPEAT(SqlFunctions.class, "repeat", String.class, int.class),
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/util/BuiltInMethod.java:343]: should this be `CHAR(SqlFunctions.class, "charN", long.class)`?

and don't you also need `CHR(SqlFunctions.class, "chr", long.class)`?

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 1 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:467]: Yes, it should be bigint.

File: core/src/main/java/org/apache/calcite/util/BuiltInMethod.java (hunk 5 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/util/BuiltInMethod.java:344]: Thank you for your careful review, I changed to them.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 2 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:668]: Added javadoc.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java (hunk 3 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:356]: Changed to it.

**Comment by k-jamroz** [comment]
Interesting changes in 1.32.0:
* https://issues.apache.org/jira/browse/CALCITE-5241 - `CHAR` is now a reserved function name. I am not sure if this may impact parsing. Maybe we want to implement this function?
* https://issues.apache.org/jira/browse/CALCITE-5232 (Upgrade protobuf-java from 3.17.1 to 3.21.5) - we currently have 3.21.9, so should be fine
* https://issues.apache.org/jira/browse/CALCITE-5178 - we currently do not support scalar queries, should not impact us

**Pull request closed without merging**